File: algorithms/77. Combinations/solution.cc
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> result;
        vector<int> path;
        bt(1, path, n, k, result);
        return result;
    }

    void bt(int start, vector<int> & path, int n, int k, vector<vector<int>> & result) {
        if ( path.size() == k ) {
            result.push_back(path);
            return;
        }
        for ( int i = start; i <= n; ++i ) {
            path.push_back(i);
            bt(i + 1, path, n, k, result);
            path.pop_back();
        }
    }
};

File: algorithms/77. Combinations/solution.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        def dfs(start, path):
            if len(path) == k:
                result.append(path[:])
                return
            for i in xrange(start, n + 1):
                path.append(i)
                dfs(i + 1, path)
                path.pop();
        result = []
        dfs(1, [])
        return result
